File: sync/js/sales_invoice.js
frappe.ui.form.on('Sales Invoice', {
    refresh(frm) {
        frm.add_custom_button(__('Send via WhatsApp'), function () {
            frappe.call({
                // method: "your_app_path.whatsapp.send_invoice_whatsapp_button",
                method: 'sync.whatsapp.send_invoice_whatsapp_button',
                args: {
                    docname: frm.doc.name
                },
                callback(r) {
                    if (r.message) {
                        frappe.msgprint(r.message.message || "WhatsApp response received.");
                    }
                }
            });
        });
    }
});

frappe.ui.form.on('Sales Invoice Item', {
    fetch_last_rate: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];

        if (!frm.doc.customer || !row.item_code) {
            frappe.msgprint("Please select both Customer and Item.");
            return;
        }

        frappe.call({
            method: "sync.sales_invoice.get_last_rates",
            args: {
                customer: frm.doc.customer,
                item_code: row.item_code
            },
            callback: function (r) {
                if (r.message) {
                    const { selling_rate, purchase_rate } = r.message;

                    let msg = "";
                    if (selling_rate) {
                        msg += "💰 Last Selling Rate: " + selling_rate + "<br>";
                    } else {
                        msg += "No previous selling rate found.<br>";
                    }

                    if (purchase_rate) {
                        msg += "🛒 Last Purchase Rate: " + purchase_rate;
                    } else {
                        msg += "No previous purchase rate found.";
                    }

                    frappe.msgprint(msg);
                } else {
                    frappe.msgprint("No rate data found.");
                }
            }
        });
    }
});

frappe.ui.form.on('Sales Invoice Item', {
    rate: function(frm, cdt, cdn) {
        update_include_gst_rate(frm, cdt, cdn);
    },
    gst_rate: function(frm, cdt, cdn) {
        update_include_gst_rate(frm, cdt, cdn);
        update_base_rate_from_include(frm, cdt, cdn);
    },
    including_gst_rate: function(frm, cdt, cdn) {
        update_base_rate_from_include(frm, cdt, cdn);
    },
    items_add: function(frm, cdt, cdn) {
        update_include_gst_rate(frm, cdt, cdn);
    }
});

function update_include_gst_rate(frm, cdt, cdn) {
    let row = locals[cdt][cdn];
    if (row.rate && row.gst_rate != null) {
        row.including_gst_rate = row.rate * (1 + row.gst_rate / 100);
        frm.refresh_field('items');
    }
}

function update_base_rate_from_include(frm, cdt, cdn) {
    let row = locals[cdt][cdn];
    if (row.including_gst_rate && row.gst_rate != null) {
        row.rate = row.including_gst_rate / (1 + row.gst_rate / 100);
        frm.refresh_field('items');
    }
}

frappe.ui.form.on('Sales Invoice', {
    margin_per: function(frm) {
      calculate_margin_total(frm);
    },
  
    grand_total: function(frm) {
      calculate_margin_total(frm);
    }
});

function calculate_margin_total(frm) {
    if (frm.doc.margin_per && frm.doc.total) {
      frm.set_value(
        'margin_amount',
        (frm.doc.margin_per / 100) * frm.doc.total
      );
    }
}
  
// OTP when Customer is overdue 

frappe.ui.form.on('Sales Invoice', {
    validate: function(frm) {
        if (!frm.doc.customer) return;

        // If already bypassed, allow save
        if (frm.doc.by_pass) {
            return;
        }

        frappe.call({
            method: "frappe.client.get_list",
            args: {
                doctype: "Sales Invoice",
                filters: {
                    customer: frm.doc.customer,
                    docstatus: 1,
                    status: "Overdue"
                },
                fields: ["name"]
            },
            async: false,
            callback: function(r) {
                if (r.message && r.message.length > 0) {
                    // Overdue found → send OTP
                    frappe.call({
                        method: "sync.sales_invoice.send_overdue_otp",
                        args: { customer: frm.doc.customer },
                        async: false,
                        callback: function(res) {
                            if (res.message && res.message.success) {
                                frappe.prompt(
                                    [
                                        {
                                            label: 'Enter OTP Wich iS sent on 9890472868',
                                            fieldname: 'otp',
                                            fieldtype: 'Data',
                                            reqd: true
                                        }
                                    ],
                                    function(values) {
                                        frappe.call({
                                            method: "sync.sales_invoice.verify_overdue_otp",
                                            args: {
                                                customer: frm.doc.customer,
                                                otp: values.otp
                                            },
                                            async: false,
                                            callback: function(vr) {
                                                if (vr.message && vr.message.verified) {
                                                    frm.set_value("by_pass", 1);
                                                    frappe.msgprint("✅ OTP Verified. Please try saving again.");
                                                    frappe.validated = false; // block current save
                                                } else {
                                                    frappe.msgprint("❌ Invalid OTP. Cannot save.");
                                                    frappe.validated = false;
                                                }
                                            }
                                        });
                                    },
                                    'OTP Verification',
                                    'Verify'
                                );

                                frappe.validated = false; // block first save until OTP handled
                            } else {
                                frappe.msgprint("❌ Failed to send OTP. Cannot save.");
                                frappe.validated = false;
                            }
                        }
                    });
                }
            }
        });
    }
});

// New-1

frappe.ui.form.on("Sales Invoice", {

    // ========================================================
    // REFRESH
    // ========================================================

    refresh(frm) {

        // MOP in Sales Invoice Item is fetched from Item Master.
        // Users should not manually edit it.
        if (
            frm.fields_dict.items &&
            frm.fields_dict.items.grid
        ) {

            frm.fields_dict.items.grid.update_docfield_property(
                "mop",
                "read_only",
                1
            );
        }
    },


    // ========================================================
    // BEFORE SAVE
    // ========================================================

    before_save: async function(frm) {

        // ----------------------------------------------------
        // Get items from Sales Invoice
        // ----------------------------------------------------

        const items = [];

        (frm.doc.items || []).forEach(row => {

            if (!row.item_code) {
                return;
            }

            items.push({
                idx: row.idx,
                item_code: row.item_code,
                rate: flt(row.rate)
            });
        });


        // ----------------------------------------------------
        // Nothing to check
        // ----------------------------------------------------

        if (!items.length) {
            return;
        }


        // ----------------------------------------------------
        // Check whether OTP approval already exists
        // ----------------------------------------------------

        const check_response = await frappe.call({

            method:
                "sync.mop_otp.check_mop_approval",

            args: {
                items: JSON.stringify(items)
            },

            freeze: true,

            freeze_message:
                "Checking MOP approval..."
        });


        const check_result =
            check_response.message || {};


        // ----------------------------------------------------
        // No MOP violation OR already approved
        // ----------------------------------------------------

        if (
            check_result.approved
        ) {
            return;
        }


        // ----------------------------------------------------
        // OTP REQUIRED
        // ----------------------------------------------------

        const verified =
            await show_mop_otp_dialog(
                frm,
                items
            );


        // ----------------------------------------------------
        // OTP failed/cancelled
        // ----------------------------------------------------

        if (!verified) {

            frappe.validated = false;

            frappe.throw(
                "Sales Invoice cannot be saved without MOP OTP verification."
            );

            return;
        }


        // ----------------------------------------------------
        // OTP verified
        //
        // IMPORTANT:
        // We do NOT call frm.save() again here.
        //
        // The current save operation continues after
        // before_save completes.
        // ----------------------------------------------------

        return;
    }

});


// ============================================================
// MOP OTP DIALOG
// ============================================================

async function show_mop_otp_dialog(frm, items) {

    return new Promise(async (resolve) => {

        let resolved = false;

        const resolve_once = function(value) {

            if (resolved) {
                return;
            }

            resolved = true;

            resolve(value);
        };


        // ----------------------------------------------------
        // Build MOP item display
        // ----------------------------------------------------

        const display_items = [];

        (items || []).forEach(row => {

            const invoice_row =
                (frm.doc.items || []).find(
                    r => r.idx === row.idx
                );

            if (!invoice_row) {
                return;
            }

            const mop = flt(invoice_row.mop);
            const rate = flt(invoice_row.rate);

            if (
                mop > 0 &&
                rate < mop
            ) {

                display_items.push({
                    idx: row.idx,
                    item_code: row.item_code,
                    rate: rate,
                    mop: mop
                });
            }
        });


        // ----------------------------------------------------
        // HTML
        // ----------------------------------------------------

        const item_html =
            display_items.map(row => {

                return `
                    <tr>
                        <td>
                            ${frappe.utils.escape_html(
                                row.item_code
                            )}
                        </td>

                        <td style="text-align:right;">
                            ${format_currency(row.rate)}
                        </td>

                        <td style="text-align:right;">
                            ${format_currency(row.mop)}
                        </td>
                    </tr>
                `;

            }).join("");


        // ----------------------------------------------------
        // Dialog
        // ----------------------------------------------------

        const d = new frappe.ui.Dialog({

            title: "MOP Approval Required",

            fields: [

                {
                    fieldname: "info",
                    fieldtype: "HTML",

                    options: `
                        <div class="alert alert-warning">
                            <strong>
                                MOP Approval Required
                            </strong>

                            <br>

                            One or more item rates are below
                            the Minimum Offer Price (MOP).
                        </div>

                        <table class="table table-bordered">

                            <thead>
                                <tr>
                                    <th>Item</th>
                                    <th style="text-align:right;">
                                        Rate
                                    </th>
                                    <th style="text-align:right;">
                                        MOP
                                    </th>
                                </tr>
                            </thead>

                            <tbody>
                                ${item_html}
                            </tbody>

                        </table>

                        <div class="alert alert-info">
                            OTP will be sent to the authorized
                            mobile number.
                        </div>
                    `
                },

                {
                    fieldname: "otp",
                    fieldtype: "Data",
                    label: "Enter OTP",

                    description:
                        "OTP is valid for 5 minutes."
                }

            ],

            primary_action_label:
                "Verify OTP",

            primary_action:
                async function() {

                    const otp =
                        d.get_value("otp");


                    // ------------------------------------------------
                    // Validate OTP input
                    // ------------------------------------------------

                    if (!otp) {

                        frappe.msgprint({
                            title: "OTP Required",
                            message:
                                "Please enter the OTP.",
                            indicator: "orange"
                        });

                        return;
                    }


                    // ------------------------------------------------
                    // Verify OTP
                    // ------------------------------------------------

                    const response =
                        await frappe.call({

                            method:
                                "sync.mop_otp.verify_mop_otp",

                            args: {

                                token:
                                    frm.__mop_otp_token,

                                otp: otp,

                                items:
                                    JSON.stringify(items)
                            },

                            freeze: true,

                            freeze_message:
                                "Verifying OTP..."
                        });


                    const result =
                        response.message || {};


                    // ------------------------------------------------
                    // Success
                    // ------------------------------------------------

                    if (
                        result.verified
                    ) {

                        d.hide();

                        frappe.show_alert({
                            message:
                                "MOP OTP verified successfully.",
                            indicator: "green"
                        });

                        resolve_once(true);

                        return;
                    }


                    // ------------------------------------------------
                    // Failed
                    // ------------------------------------------------

                    frappe.msgprint({

                        title:
                            "OTP Verification Failed",

                        message:
                            result.message ||
                            "Invalid or expired OTP.",

                        indicator:
                            "red"
                    });
                },


            secondary_action_label:
                "Cancel",

            secondary_action:
                function() {

                    d.hide();

                    resolve_once(false);
                }
        });


        // ----------------------------------------------------
        // Show dialog
        // ----------------------------------------------------

        d.show();


        // ----------------------------------------------------
        // Generate temporary browser token
        //
        // This is NOT a Sales Invoice field.
        // It exists only in JavaScript memory.
        // ----------------------------------------------------

        frm.__mop_otp_token =
            frappe.utils.get_random(32);


        // ----------------------------------------------------
        // Send OTP automatically
        // ----------------------------------------------------

        const send_response =
            await frappe.call({

                method:
                    "sync.mop_otp.send_mop_otp",

                args: {

                    token:
                        frm.__mop_otp_token,

                    items:
                        JSON.stringify(items)
                },

                freeze: true,

                freeze_message:
                    "Sending MOP OTP..."
            });


        const send_result =
            send_response.message || {};


        // ----------------------------------------------------
        // SMS sent
        // ----------------------------------------------------

        if (
            send_result.success
        ) {

            frappe.show_alert({

                message:
                    "OTP sent successfully. Valid for 5 minutes.",

                indicator:
                    "green"
            });

            return;
        }


        // ----------------------------------------------------
        // SMS failed
        // ----------------------------------------------------

        d.hide();

        frappe.msgprint({

            title:
                "OTP Sending Failed",

            message:
                send_result.message ||
                "Unable to send OTP.",

            indicator:
                "red"
        });

        resolve_once(false);
    });
}

File: sync/sales_invoice.py
import frappe
import random
import requests

@frappe.whitelist()
def get_last_rates(customer, item_code):
    selling_rate = frappe.db.sql("""
        SELECT sii.rate FROM `tabSales Invoice Item` sii
        JOIN `tabSales Invoice` si ON sii.parent = si.name
        WHERE sii.item_code = %s AND si.customer = %s
        AND si.docstatus = 1
        ORDER BY si.posting_date DESC, si.modified DESC
        LIMIT 1
    """, (item_code, customer), as_dict=1)

    purchase_rate = frappe.db.sql("""
        SELECT pii.rate FROM `tabPurchase Invoice Item` pii
        JOIN `tabPurchase Invoice` pi ON pii.parent = pi.name
        WHERE pii.item_code = %s AND pi.docstatus = 1
        ORDER BY pi.posting_date DESC, pi.modified DESC
        LIMIT 1
    """, (item_code,), as_dict=1)

    return {
        "selling_rate": selling_rate[0].rate if selling_rate else None,
        "purchase_rate": purchase_rate[0].rate if purchase_rate else None
    }

## New Code 

@frappe.whitelist()
def send_overdue_otp(customer):
    phone = "8423594555"  # Must include country code if required by API

    # phone = "7977185868"
    if not phone:
        return {"success": False, "error": "No phone number found"}

    otp = str(random.randint(100000, 999999))
    # frappe.msgprint(f"Generated OTP: {otp}")

    customer_name = frappe.db.get_value("Customer", customer, "customer_name")
    # Store OTP in cache for 5 minutes
    frappe.cache().set_value(f"overdue_otp_{customer}", otp, expires_in_sec=300)

    # Send OTP
    # return send_whatsapp_message(phone, otp)
    return send_whatsapp_message(phone, otp, customer_name)


@frappe.whitelist()
def verify_overdue_otp(customer, otp):
    cached_otp = frappe.cache().get_value(f"overdue_otp_{customer}")
    if cached_otp and cached_otp == otp:
        frappe.cache().delete_value(f"overdue_otp_{customer}")
        return {"verified": True}
    return {"verified": False}


@frappe.whitelist()
# def send_whatsapp_message(receiver_mobile_no, otp):
def send_whatsapp_message(receiver_mobile_no, otp, customer_name=None):
    # url = "https://wbcapi.messagerider.com/MessageRider/SendMsg"
    url = "https://wbcapi.messagerider.com/MessageRider/SendMsg"
    
    # message = f" <b>{customer_name}<b>, Your verification OTP is {otp}" if customer_name else f"Your verification OTP is {otp}"
    
    message = f"Customer Name : {customer_name}, Your verification OTP is {otp}" if customer_name else f"Your verification OTP is {otp}"
    params = {
        "authtoken": "0000146",
        "password": "Prism@2025",
        # "message": f"Your verification OTP is {otp}",
        "message": message,
        "receiverMobileNo": receiver_mobile_no
    }
    
    try:
        response = requests.get(url, params=params, timeout=10)
        
        if response.status_code == 200:
            return {
                "success": True,
                "response": response.text
            }
        else:
            frappe.throw(f"Error from API: {response.status_code} - {response.text}")
    except Exception as e:
        frappe.throw(f"API Request Failed: {str(e)}")

# MOP Code

from sync.mop_otp import (
    get_mop_items,
    is_mop_approval_valid,
)


def validate_mop(doc, method=None):
    """
    Server-side MOP validation.

    Rules:
    - Rate >= MOP -> normal save/submit
    - Rate < MOP -> valid OTP approval required

    MOP is always fetched from Item Master on the server.
    No Sales Invoice field is used for OTP/token storage.
    """

    mop_items = get_mop_items(doc)

    # No MOP violation
    if not mop_items:
        return

    # Check whether OTP approval exists for the exact
    # current item/rate/MOP combination.
    if is_mop_approval_valid(mop_items):
        return

    frappe.throw(
        "MOP approval is required because one or more item "
        "rates are below the Minimum Offer Price (MOP). "
        "Please verify the OTP."
    )

# # import frappe

# from sync.mop_otp import (
#     get_mop_items,
#     consume_mop_verification
# )


# def validate_mop(doc, method=None):
#     """
#     Server-side MOP validation for Sales Invoice.

#     If Rate < MOP, valid OTP verification is required.
#     """

#     mop_items = get_mop_items(doc)

#     # No MOP violation
#     if not mop_items:
#         return

#     token = doc.get("custom_mop_otp_token")

#     if not token:
#         frappe.throw(
#             "MOP approval is required because one or more item rates "
#             "are below the Minimum Offer Price."
#         )

#     # Verify OTP approval
#     verified = consume_mop_verification(
#         doc.name,
#         token
#     )

#     if not verified:

#         frappe.throw(
#             "MOP OTP verification is required before saving/submitting "
#             "this Sales Invoice."
#         )